accord-core/src/main/java/accord/local/Command.java

```diff
@@ -592,75 +839,320 @@ public Key someKey()
+                    throw new IllegalStateException();
+            }
+        }
+
+        // invalid to Add deps to Accepted or AcceptedInvalidate statuses, as Committed deps are not equivalent
+        // and we may erroneously believe we have covered a wider range than we have infact covered
+        if (ensurePartialDeps == Add)
+            Preconditions.checkState(status() != Accepted && status() != AcceptedInvalidate);
+
+        // validate new partial txn
+        if (!validate(ensurePartialTxn, existingRanges, additionalRanges, covers(partialTxn()), covers(partialTxn), "txn", partialTxn))
+            return false;
+
+        if (shard.isHome() && ensurePartialTxn != Ignore)
+        {
+            if (!hasQuery(partialTxn()) && !hasQuery(partialTxn))
+                throw new IllegalStateException();
+        }
+
+        return validate(ensurePartialDeps, existingRanges, additionalRanges, covers(partialDeps()), covers(partialDeps), "deps", partialDeps);
     }
 
-    private PartialCommand directlyBlockedBy()
+    private void set(SafeCommandStore safeStore,
+                     KeyRanges existingRanges, KeyRanges additionalRanges, ProgressShard shard, AbstractRoute route,
+                     @Nullable PartialTxn partialTxn, EnsureAction ensurePartialTxn,
+                     @Nullable PartialDeps partialDeps, EnsureAction ensurePartialDeps)
     {
-        // firstly we're waiting on every dep to commit
-        while (isWaitingOnCommit())
+        Preconditions.checkState(progressKey() != null);
+        KeyRanges allRanges = existingRanges.union(additionalRanges);
+
+        if (shard.isProgress()) setRoute(AbstractRoute.merge(route(), route));
+        else setRoute(AbstractRoute.merge(route(), route.slice(allRanges)));
+
+        // TODO (soon): stop round-robin hashing; partition only on ranges
+        switch (ensurePartialTxn)
         {
-            // TODO: when we change our liveness mechanism this may not be a problem
-            // cannot guarantee that listener updating this set is invoked before this method by another listener
-            // so we must check the entry is still valid, and potentially remove it if not
-            PartialCommand waitingOn = firstWaitingOnCommit();
-            if (!waitingOn.hasBeen(Committed)) return waitingOn;
-            onChangeInternal(waitingOn);
+            case Add:
+                if (partialTxn == null)
+                    break;
+
+                if (partialTxn() != null)
+                {
+                    partialTxn = partialTxn.slice(allRanges, shard.isHome());
+                    partialTxn.keys().foldlDifference(partialTxn().keys(), (i, key, p, v) -> {
+                        if (safeStore.commandStore().hashIntersects(key))
+                            safeStore.commandsForKey(key).register(this);
+                        return v;
+                    }, 0, 0, 1);
+                    this.setPartialTxn(partialTxn().with(partialTxn));
```

Comment:
Why would we want with and not also get the null checks that come with merge?

Reply:
I think this will all likely be deprecated by the work Blake is doing, so I wouldn't agonise over it.